File: src/codex_subagent_kit/doctor.py
from __future__ import annotations

import tomllib
from dataclasses import dataclass
from pathlib import Path

from .catalog import (
    BUILTIN_CATEGORIES_DIR,
    IMPORTED_AGENTS_CATEGORY,
    _parse_agent_file,
    _parse_category_dir,
    normalize_catalog_roots,
    resolve_global_agents_dir,
    resolve_global_catalog_dir,
    resolve_project_agents_dir,
    resolve_project_catalog_dir,
)
from .generator import resolve_target_dir
from .models import AgentSpec, Category


@dataclass(frozen=True)
class DoctorIssue:
    path: Path | None
    message: str


@dataclass(frozen=True)
class DoctorReport:
    scope: str
    target_dir: Path
    catalog_counts: list[tuple[str, int]]
    installed_counts: list[tuple[str, int]]
    issues: list[DoctorIssue]

    @property
    def ok(self) -> bool:
        return not self.issues


def _scan_catalog_root(
    root: Path,
    *,
    source: str,
    missing_is_issue: bool = False,
) -> tuple[dict[str, Category], dict[str, AgentSpec], list[DoctorIssue], int]:
    if not root.exists():
        issues = [DoctorIssue(root, "catalog root does not exist")] if missing_is_issue else []
        return {}, {}, issues, 0

    categories: dict[str, Category] = {}
    agents: dict[str, AgentSpec] = {}
    issues: list[DoctorIssue] = []
    checked_templates = 0

    for category_dir in sorted(root.iterdir()):
        if not category_dir.is_dir():
            continue
        category = _parse_category_dir(category_dir)
        categories[category.key] = category
        for path in sorted(category_dir.glob("*.toml")):
            checked_templates += 1
            try:
                agent = _parse_agent_file(path, inherited_category=category.key, source=source)
            except (OSError, ValueError, tomllib.TOMLDecodeError) as exc:
                issues.append(DoctorIssue(path, str(exc)))
                continue
            agents[agent.key] = agent

    return categories, agents, issues, checked_templates


def _scan_installed_agents(
    directory: Path,
    *,
    source: str,
    inherited_agents: dict[str, AgentSpec],
) -> tuple[list[AgentSpec], list[DoctorIssue], int]:
    if not directory.exists():
        return [], [], 0

    agents: list[AgentSpec] = []
    issues: list[DoctorIssue] = []
    checked_files = 0

    for path in sorted(directory.glob("*.toml")):
        checked_files += 1
        inherited_category = inherited_agents.get(path.stem).category if path.stem in inherited_agents else None
        try:
            agent = _parse_agent_file(path, inherited_category=inherited_category, source=source)
        except (OSError, ValueError, tomllib.TOMLDecodeError) as exc:
            issues.append(DoctorIssue(path, str(exc)))
            continue
        agents.append(agent)

    return agents, issues, checked_files


def run_doctor(
    *,
    project_root: Path,
    scope: str,
    catalog_roots: tuple[Path, ...] = (),
) -> DoctorReport:
    normalized_catalog_roots = normalize_catalog_roots(catalog_roots)

    catalog_counts: list[tuple[str, int]] = []
    installed_counts: list[tuple[str, int]] = []
    issues: list[DoctorIssue] = []

    categories, agent_map, builtin_issues, builtin_checked = _scan_catalog_root(
        BUILTIN_CATEGORIES_DIR,
        source="builtin",
    )
    issues.extend(builtin_issues)
    catalog_counts.append(("built-in", builtin_checked))

    if scope == "project":
        global_categories, global_agents, global_issues, global_checked = _scan_catalog_root(
            resolve_global_catalog_dir(),
            source="global-catalog",
        )
        categories.update(global_categories)
        agent_map.update(global_agents)
        issues.extend(global_issues)
        catalog_counts.append(("global catalog", global_checked))

        project_categories, project_agents, project_issues, project_checked = _scan_catalog_root(
            resolve_project_catalog_dir(project_root),
            source="project-catalog",
        )
        categories.update(project_categories)
        agent_map.update(project_agents)
        issues.extend(project_issues)
        catalog_counts.append(("project catalog", project_checked))
    else:
        global_categories, global_agents, global_issues, global_checked = _scan_catalog_root(
            resolve_global_catalog_dir(),
            source="global-catalog",
        )
        categories.update(global_categories)
        agent_map.update(global_agents)
        issues.extend(global_issues)
        catalog_counts.append(("global catalog", global_checked))

    for root in normalized_catalog_roots:
        extra_categories, extra_agents, extra_issues, extra_checked = _scan_catalog_root(
            root,
            source="catalog-root",
            missing_is_issue=True,
        )
        categories.update(extra_categories)
        agent_map.update(extra_agents)
        issues.extend(extra_issues)
        catalog_counts.append((f"catalog root: {root}", extra_checked))

    categories[IMPORTED_AGENTS_CATEGORY.key] = IMPORTED_AGENTS_CATEGORY
    del categories

    global_installed, global_installed_issues, global_installed_checked = _scan_installed_agents(
        resolve_global_agents_dir(),
        source="global",
        inherited_agents=agent_map,
    )
    issues.extend(global_installed_issues)

    if scope == "project":
        installed_counts.append(("global agents", global_installed_checked))
        for agent in global_installed:
            agent_map[agent.key] = agent

        project_installed_dir = resolve_project_agents_dir(project_root)
        project_installed, project_installed_issues, project_installed_checked = _scan_installed_agents(
            project_installed_dir,
            source="project",
            inherited_agents=agent_map,
        )
        issues.extend(project_installed_issues)
        installed_counts.append(("project agents", project_installed_checked))

        target_dir = resolve_target_dir("project", project_root)
        if project_installed_checked == 0:
            issues.append(DoctorIssue(target_dir, "no installed agent definitions found in target scope"))
    else:
        installed_counts.append(("global agents", global_installed_checked))
        target_dir = resolve_target_dir("global", project_root)
        if global_installed_checked == 0:
            issues.append(DoctorIssue(target_dir, "no installed agent definitions found in target scope"))

    return DoctorReport(
        scope=scope,
        target_dir=target_dir,
        catalog_counts=catalog_counts,
        installed_counts=installed_counts,
        issues=issues,
    )


def render_doctor_report(report: DoctorReport) -> str:
    lines = [
        f"status: {'ok' if report.ok else 'issues found'}",
        f"scope: {report.scope}",
        f"target: {report.target_dir}",
        "",
        "Catalog templates checked:",
    ]
    for label, count in report.catalog_counts:
        lines.append(f"- {label}: {count}")

    lines.append("")
    lines.append("Installed agent files checked:")
    for label, count in report.installed_counts:
        lines.append(f"- {label}: {count}")

    lines.append("")
    if report.issues:
        lines.append("Issues:")
        for issue in report.issues:
            if issue.path is None:
                lines.append(f"- {issue.message}")
            else:
                lines.append(f"- {issue.path}: {issue.message}")
    else:
        lines.append("Issues: none")

    return "\n".join(lines)

File: packages/codex-subagent-kit/src/doctor.ts
import { existsSync, readdirSync } from "node:fs";
import { resolve } from "node:path";

import {
  BUILTIN_CATEGORIES_DIR,
  getAgentMap,
  IMPORTED_AGENTS_CATEGORY,
  parseAgentFile,
  parseCategoryDir,
} from "./catalog";
import type { AgentSpec, Category, DoctorIssue, DoctorReport } from "./models";
import {
  normalizeCatalogRoots,
  resolveGlobalAgentsDir,
  resolveGlobalCatalogDir,
  resolveProjectCatalogDir,
} from "./paths";
import { resolveTargetDir } from "./generator";

function scanCatalogRoot(
  root: string,
  source: string,
  missingIsIssue = false,
): {
  categories: Map<string, Category>;
  agents: Map<string, AgentSpec>;
  issues: DoctorIssue[];
  checkedTemplates: number;
} {
  const categories = new Map<string, Category>();
  const agents = new Map<string, AgentSpec>();
  const issues: DoctorIssue[] = [];
  let checkedTemplates = 0;

  if (!existsSync(root)) {
    if (missingIsIssue) {
      issues.push({ path: root, message: "catalog root does not exist" });
    }
    return { categories, agents, issues, checkedTemplates };
  }

  for (const entry of readdirSync(root, { withFileTypes: true })) {
    if (!entry.isDirectory()) {
      continue;
    }
    const categoryDir = resolve(root, entry.name);
    const category = parseCategoryDir(categoryDir);
    categories.set(category.key, category);

    for (const file of readdirSync(categoryDir, { withFileTypes: true })) {
      if (!file.isFile() || !file.name.endsWith(".toml")) {
        continue;
      }
      const path = resolve(categoryDir, file.name);
      checkedTemplates += 1;
      try {
        const agent = parseAgentFile(path, source, category.key);
        agents.set(agent.key, agent);
      } catch (error) {
        issues.push({ path, message: error instanceof Error ? error.message : String(error) });
      }
    }
  }

  return { categories, agents, issues, checkedTemplates };
}

function scanInstalledAgents(
  directory: string,
  source: string,
  inheritedAgents: Map<string, AgentSpec>,
): { agents: AgentSpec[]; issues: DoctorIssue[]; checkedFiles: number } {
  const agents: AgentSpec[] = [];
  const issues: DoctorIssue[] = [];
  let checkedFiles = 0;

  if (!existsSync(directory)) {
    return { agents, issues, checkedFiles };
  }

  for (const file of readdirSync(directory, { withFileTypes: true })) {
    if (!file.isFile() || !file.name.endsWith(".toml")) {
      continue;
    }
    const path = resolve(directory, file.name);
    checkedFiles += 1;
    try {
      const inheritedCategory = inheritedAgents.get(file.name.replace(/\.toml$/, ""))?.category;
      agents.push(parseAgentFile(path, source, inheritedCategory));
    } catch (error) {
      issues.push({ path, message: error instanceof Error ? error.message : String(error) });
    }
  }

  return { agents, issues, checkedFiles };
}

export function runDoctor(options: {
  projectRoot: string;
  scope: "project" | "global";
  homeDir?: string;
  catalogRoots?: readonly string[];
}): DoctorReport {
  const catalogCounts: Array<[string, number]> = [];
  const installedCounts: Array<[string, number]> = [];
  const issues: DoctorIssue[] = [];

  const builtin = scanCatalogRoot(BUILTIN_CATEGORIES_DIR, "builtin");
  catalogCounts.push(["built-in", builtin.checkedTemplates]);
  issues.push(...builtin.issues);

  const categories = new Map(builtin.categories);
  const agentMap = new Map(builtin.agents);

  const globalCatalog = scanCatalogRoot(resolveGlobalCatalogDir(options.homeDir), "global-catalog");
  catalogCounts.push(["global catalog", globalCatalog.checkedTemplates]);
  issues.push(...globalCatalog.issues);
  for (const [key, category] of globalCatalog.categories) categories.set(key, category);
  for (const [key, agent] of globalCatalog.agents) agentMap.set(key, agent);

  if (options.scope === "project") {
    const projectCatalog = scanCatalogRoot(
      resolveProjectCatalogDir(options.projectRoot),
      "project-catalog",
    );
    catalogCounts.push(["project catalog", projectCatalog.checkedTemplates]);
    issues.push(...projectCatalog.issues);
    for (const [key, category] of projectCatalog.categories) categories.set(key, category);
    for (const [key, agent] of projectCatalog.agents) agentMap.set(key, agent);
  }

  for (const root of normalizeCatalogRoots(options.catalogRoots)) {
    const extra = scanCatalogRoot(root, "catalog-root", true);
    catalogCounts.push([`catalog root: ${root}`, extra.checkedTemplates]);
    issues.push(...extra.issues);
    for (const [key, category] of extra.categories) categories.set(key, category);
    for (const [key, agent] of extra.agents) agentMap.set(key, agent);
  }

  categories.set(IMPORTED_AGENTS_CATEGORY.key, IMPORTED_AGENTS_CATEGORY);

  const globalInstalled = scanInstalledAgents(
    resolveGlobalAgentsDir(options.homeDir),
    "global",
    agentMap,
  );
  installedCounts.push(["global agents", globalInstalled.checkedFiles]);
  issues.push(...globalInstalled.issues);
  for (const agent of globalInstalled.agents) {
    agentMap.set(agent.key, agent);
  }

  let targetDir: string;
  if (options.scope === "project") {
    const projectInstalledDir = resolveTargetDir("project", options.projectRoot);
    const projectInstalled = scanInstalledAgents(projectInstalledDir, "project", agentMap);
    installedCounts.push(["project agents", projectInstalled.checkedFiles]);
    issues.push(...projectInstalled.issues);
    targetDir = projectInstalledDir;
    if (projectInstalled.checkedFiles === 0) {
      issues.push({ path: targetDir, message: "no installed agent definitions found in target scope" });
    }
  } else {
    targetDir = resolveTargetDir("global", options.projectRoot, options.homeDir);
    if (globalInstalled.checkedFiles === 0) {
      issues.push({ path: targetDir, message: "no installed agent definitions found in target scope" });
    }
  }

  return {
    scope: options.scope,
    targetDir,
    catalogCounts,
    installedCounts,
    issues,
  };
}

export function renderDoctorReport(report: DoctorReport): string {
  const lines = [
    `status: ${report.issues.length === 0 ? "ok" : "issues found"}`,
    `scope: ${report.scope}`,
    `target: ${report.targetDir}`,
    "",
    "Catalog templates checked:",
  ];

  for (const [label, count] of report.catalogCounts) {
    lines.push(`- ${label}: ${count}`);
  }

  lines.push("");
  lines.push("Installed agent files checked:");
  for (const [label, count] of report.installedCounts) {
    lines.push(`- ${label}: ${count}`);
  }

  lines.push("");
  if (report.issues.length === 0) {
    lines.push("Issues: none");
  } else {
    lines.push("Issues:");
    for (const issue of report.issues) {
      lines.push(issue.path ? `- ${issue.path}: ${issue.message}` : `- ${issue.message}`);
    }
  }

  return lines.join("\n");
}
